### PR TITLE
Cached feeds with sqldelight

### LINKED ISSUE #58: Use SQLDelight to cache feeds
## Kind (Required)

- Improvement

 ## Overview (Required)

- Currently this app only holds the cache in the ViewModel. I want to cache using DB.

 ## Links	

- https://github.com/cashapp/sqldelight

### PR DESCRIPTION
**Pull request by atsuya046**
## Issue
- close #58

## Overview (Required)
- SQLDelight setup
- return cached FeedItems if these are not empty

## Links
none

## Screenshot
none

### CODE REVIEW

**Comment by atsuya046** [comment]
oops! I fix it.

**Comment by takahirom** [comment]
I think the mapping work was hard! Good job 👍 
I will see this PR tomorrow!

File: data/db/build.gradle [hunk 1]
```diff
@@ -62,6 +63,9 @@ kotlin {
             implementation Dep.Dagger.hiltAndroid
             implementation "androidx.datastore:datastore-preferences:1.0.0-alpha04"
 
+            // SQLDelight
```
Comment (takahirom) [inline data/db/build.gradle:66]: 📝 
https://cashapp.github.io/sqldelight/multiplatform_sqlite/
  // The drivers needed will change depending on what platforms you target:

File: data/db/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedItemDao.kt [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    fun insert(feeds: List<FeedItem>) {
+        feeds.forEach { item ->
+            when (item) {
+                is FeedItem.Blog -> blogQueries.insert(item)
+                is FeedItem.Podcast -> {
+                    podcastQueries.insert(item)
+                    item.speakers.forEach { speaker ->
+                        podcastSpeakerQueries.insert(item.id,
+                            speaker)
```
Comment (takahirom) [inline data/db/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedItemDao.kt:34]: Probably you can fix this by ./gradlew :data:db:spotlessKotlinApply

File: data/repository/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedRepositoryImpl.kt [hunk 3]
```diff
@@ -9,13 +9,22 @@ import kotlinx.coroutines.flow.flow
...
     private val dataStore: UserDataStore,
 ) : FeedRepository {
-    override fun feedContents(): Flow<FeedContents> {
+    override fun feedContents(forceUpdate: Boolean): Flow<FeedContents> {
         return dataStore.favorites()
             .combine(
                 flow {
-                    emit(feedApi.fetch())
+                    val cachedFeeds by lazy { feedItemDao.selectAll() }
+                    if (forceUpdate || cachedFeeds.isEmpty()) {
```
Comment (takahirom) [inline data/repository/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedRepositoryImpl.kt:20]: Ideally I would like to use the coroutines extension to use the db data as a single source of truth in the Flow, and have a separate refersh method to call it. 🙏 
https://cashapp.github.io/sqldelight/multiplatform_sqlite/coroutines/

Is it possible to do so?
If it is difficult, you can set forceUpdate to true and implement it later.

viewmodel

```kotlin
private val allFeedContents: StateFlow<LoadState<FeedContents>> = repository.feedContents()
        ...
    init {
        repository.refresh()
    }
```

repository

```kotlin
    override fun feedContents(forceUpdate: Boolean): Flow<FeedContents> {
        return dataStore.favorites()
            .combine( feedItemDao.selectAll()) // this return flow
            {  favorites, apiFeed -> ...
            }
   }
   fun refresh() {
      // fetch API and update DB
   }
```

Comment (atsuya046) [reply data/repository/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedRepositoryImpl.kt]: 534e9a8 I implemented with coroutines extension, and separate refresh method!!

File: data/db/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedItemDao.kt [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+                media = Media.parse(media),
+                title = MultiLangText(jaTitle = jaTitle, enTitle = enTitle),
+                summary = MultiLangText(jaTitle = jaSummary, enTitle = enSummary),
+                link = link
+            )
+        }
+    }
+}
+
+private fun FeedItemBlogQueries.insert(blog: FeedItem.Blog) {
```
Comment (takahirom) [inline data/db/src/commonMain/kotlin/io/github/droidkaigi/feeder/data/FeedItemDao.kt:134]: 👍

File: android/src/test/java/io/github/droidkaigi/feeder/FeedItemViewModelTest.kt [hunk 5]
```diff
@@ -116,8 +117,19 @@ class FeedItemViewModelTest(
...
+                                    null
+                                }
+                            ),
                             dataStore = fakeUserDataStore()
-                        )
+                        ).also {
+                            runBlockingTest {
+                                if (!errorFetchData) it.refresh()
+                            }
+                        }
```
Comment (takahirom) [inline android/src/test/java/io/github/droidkaigi/feeder/FeedItemViewModelTest.kt:132]: Thanks for fixing the test too!
Why is this code needed? I tried erasing it and moving it, but the test seems to pass 👀

**Review by takahirom — COMMENTED** [review]
Almost LGTM 🥇 (Please check comment)

File: android/src/test/java/io/github/droidkaigi/feeder/FeedItemViewModelTest.kt (hunk 5 above)
Comment (atsuya046) [reply android/src/test/java/io/github/droidkaigi/feeder/FeedItemViewModelTest.kt]: Sorry, you are right 😅 
I forgot refreshing on ViewModel init.
I removed this 🙏

**Review by takahirom — APPROVED** [review]
I didn't know about sqldelight! I learned a lot! LGTM!!

**Pull request merged**